x/wasm/keeper/keeper.go

```diff
@@ -1213,3 +1218,20 @@ func (h DefaultWasmVMContractResponseHandler) Handle(ctx sdk.Context, contractAd
 	}
 	return result, nil
 }
+
+// PruneWasmCodes deletes code info for unpinned codes.
+func (k Keeper) PruneWasmCodes(ctx sdk.Context, codeIDs []uint64) error {
+	for _, c := range codeIDs {
+		if k.IsPinnedCode(ctx, c) {
+			continue
+		}
+
+		info := k.GetCodeInfo(ctx, c)
+		if info == nil {
+			return types.ErrNoSuchCodeFn(c).Wrapf("code id %d", c)
+		}
+		k.wasmVM.RemoveCode(info.CodeHash)
```

Comment:
ah, we are removing code by checksum and not codeID. It is very likely that multiple codeIDs for the same hash exist. We need to ensure that we only delete by a checksum when no instance for any of codeID exists.